File: Flutter/voiceroomkit/voiceroomkit/lib/src/api/model/nevoiceroom_member.dart
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

part of netease_voiceroomkit;

/// 语聊房成员对象
class NEVoiceRoomMember {
  /// 用户id
  final String account;

  /// 用户名
  final String name;

  /// 用户角色
  final String role;

  /// 音频是否打开
  final bool isAudioOn;

  /// 音频是否被禁用
  final bool isAudioBanned;

  /// 头像
  final String? avatar;

  NEVoiceRoomMember(this.account, this.name, this.role, this.isAudioOn,
      this.isAudioBanned, this.avatar);

  @override
  String toString() {
    return 'NEVoiceRoomMember{account: $account, name: $name, role: $role, isAudioOn: $isAudioOn, isAudioBanned: $isAudioBanned, avatar: $avatar}';
  }
}

File: iOS/VoiceRoomKit/NEVoiceRoomKit/NEVoiceRoomKit/Classes/Public/Common/NEVoiceRoomMember.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NERoomKit

/// 成员属性变量
struct MemberPropertyConstants {
  enum MuteAudio {
    static let key = "recordDevice"
    static let on = "on"
    static let off = "off"
  }

  enum CanOpenMic {
    static let key = "canOpenMic"
    static let no = "0"
    static let yes = "1"
  }
}

@objcMembers
/// 用户音量信息
public class NEVoiceRoomMemberVolumeInfo: NSObject {
  /// 成员ID
  public var userUuid: String = ""
  /// 音量大小 区间  **[0~100]**
  public var volume: Int = 0

  init(info: NEMemberVolumeInfo) {
    userUuid = info.userUuid
    volume = info.volume
  }
}

@objcMembers
/// Karaoke 成员模型
public class NEVoiceRoomMember: NSObject {
  /// 用户ID
  public var account: String = ""
  /// 用户名
  public var name: String = ""
  /// 用户角色
  public var role: String = ""
  /// 用户头像
  public var avatar: String?
  /// 音频是否打开
  public var isAudioOn: Bool = false
  /// 音频是否被禁用
  public var isAudioBanned: Bool = false
  override public init() {
    super.init()
  }

  convenience init(_ member: NERoomMember) {
    self.init()
    account = member.uuid
    name = member.name
    role = member.role.name
    avatar = member.avatar

    isAudioOn = member.properties[MemberPropertyConstants.MuteAudio.key] == MemberPropertyConstants
      .MuteAudio.on
    isAudioBanned = member
      .properties[MemberPropertyConstants.CanOpenMic.key] == MemberPropertyConstants.CanOpenMic.no
  }

  init?(member: NERoomMember?) {
    guard let member = member else {
      return nil
    }
    account = member.uuid
    name = member.name
    role = member.role.name
    avatar = member.avatar

    isAudioOn = member.properties[MemberPropertyConstants.MuteAudio.key] == MemberPropertyConstants
      .MuteAudio.on
    isAudioBanned = member
      .properties[MemberPropertyConstants.CanOpenMic.key] == MemberPropertyConstants.CanOpenMic.no
  }
}
